firedrake/pointquery_utils.py

```diff
@@ -286,90 +74,84 @@ def X_isub_dX(topological_dimension):
-        # if needs_orientation:
-        #     code_ += [format["orientation"]["ufc"](tdim, gdim)]
-
-        x = np.array([sp.Symbol("x[%d]" % i) for i in range(gdim)])
-        x0 = np.array([sp.Symbol("x0[%d]" % i) for i in range(gdim)])
-        K = np.array([[sp.Symbol("K[%d]" % (i*gdim + j)) for j in range(gdim)]
-                      for i in range(tdim)])
-
-        dX = K.dot(x - x0)
-        for i, e in enumerate(dX):
-            code += ["\tdX[%d] = %s;" % (i, e)]
-
-        return "\n".join(code)
-
-    # Create FIAT element
-    element = create_element(ufl_coordinate_element, vector_is_mixed=False)
-    cell = ufl_coordinate_element.cell()
+    def init_X(fiat_cell):
+        vertices = numpy.array(fiat_cell.get_vertices())
+        X = numpy.average(vertices, axis=0)
+
+        formatter = ArrayInit(X, precision=parameters["precision"])._formatter
+        return "\n".join("%s = %s;" % ("X[%d]" % i, formatter(v)) for i, v in enumerate(X))
+
+    def to_reference_coordinates(ufl_coordinate_element):
+        # Set up UFL form
+        cell = ufl_coordinate_element.cell()
+        domain = ufl.Mesh(ufl_coordinate_element)
+        K = ufl.JacobianInverse(domain)
+        x = ufl.SpatialCoordinate(domain)
+        x0_element = ufl.VectorElement("Real", cell, 0)
+        x0 = ufl.Coefficient(ufl.FunctionSpace(domain, x0_element))
+        expr = ufl.dot(K, x - x0)
+
+        # Translation to GEM
+        C = ufl_utils.coordinate_coefficient(domain)
+        expr = ufl_utils.preprocess_expression(expr)
+        expr = ufl_utils.replace_coordinates(expr, C)
+        expr = ufl_utils.simplify_abs(expr)
+
+        builder = firedrake_interface.KernelBuilderBase()
+        builder._coefficient(C, "C")
+        builder._coefficient(x0, "x0")
+
+        dim = cell.topological_dimension()
+        point = gem.Variable('X', (dim,))
+        context = tsfc.fem.GemPointContext(
+            interface=builder,
+            ufl_cell=cell,
+            precision=parameters["precision"],
+            point_indices=(),
+            point_expr=point,
+        )
+        translator = tsfc.fem.Translator(context)
+        ir = map_expr_dag(translator, expr)
+
+        # Unroll result
+        ir = [gem.Indexed(ir, alpha) for alpha in numpy.ndindex(ir.shape)]
+
+        # Unroll IndexSums
+        max_extent = parameters["unroll_indexsum"]
+        if max_extent:
+            def predicate(index):
+                return index.extent <= max_extent
+        ir = gem.optimise.unroll_indexsum(ir, predicate=predicate)
+
+        # Translate to COFFEE
+        ir = impero_utils.preprocess_gem(ir)
+        return_variable = gem.Variable('dX', (dim,))
+        assignments = [(gem.Indexed(return_variable, (i,)), e)
+                       for i, e in enumerate(ir)]
+        impero_c = impero_utils.compile_gem(assignments, ())
+        body = tsfc.coffee.generate(impero_c, {}, parameters["precision"])
+        body.open_scope = False
+
+        return body
```

Comment:
This is the third place I see something like this, which suggests a potential refactoring opportunity at some later occasion...